src/hardware/sblaster.cpp

```diff
@@ -2850,125 +2981,44 @@ bool SB_GetAddress(uint16_t &sbaddr, uint8_t &sbirq, uint8_t &sbdma)
 	return (sbaddr != 0 && sbirq != 0 && sbdma != 0);
 }
 
-static void sblaster_mixer_callback([[maybe_unused]]const int requested_frames)
+static void generate_frame()
 {
-	// We can ignore requested frames as this function gets called in a loop until it gets what it needs
-	// Overflow is not a concern as extra frames will remain in the channel's buffer and get mixed on the next callback
-
-	const auto frames = soundblaster_mixer_queue.Dequeue();
-	if (!frames) {
-		// Queue must be stopped, otherwise Dequeue() will block until some frames are available
-		sb.chan->AddSilence();
-		return;
-	}
-	// Weird double de-referencing syntax as the object type is std::optional<std::unique_ptr<AudioVector>>
-	(*frames)->AddSamples(sb.chan.get());
-}
-
-static void generate_frames(const int frames_requested)
-{
-	static int ticks_of_silence = 0;
+	assert(sblaster && sblaster->output_queue.IsRunning());
 
 	switch (sb.mode) {
 	case DspMode::None:
 	case DspMode::DmaPause:
 	case DspMode::DmaMasked:
-	if (!soundblaster_mixer_queue.IsRunning()) {
-		// We've been playing silence and are continuing to play silence
-		// Nothing to do except increment this variable so we don't loop forever
-		frames_added_this_tick += frames_requested;
-		break;
-	}
-	// Enqueue a tick's worth of silenced frames
-	// Some games (Tyrian for example) will switch to DmaMasked briefly (less than 5 ticks usually)
-	// We can't use AddSilence on the mixer thread as that asks for a blocksize of audio (usually over 10ms)
-	enqueue_frames(std::make_unique<AudioVectorM8S>(
-		frames_requested,
-		std::vector<int8_t>(frames_requested).data()
-	));
-
-	++ticks_of_silence;
-	if (ticks_of_silence > 5000) {
-		// We've been playing silence for 5 seconds
-		// Some games only play DMA sound in certain segments or for small durations
-		// Either we're not in a DMA game at all or its been quiet for a while
-		// Stop the mixer channel for performance reasons as this channel blocks waiting for the main thread to provide more audio
-		// 5 seconds is safe to avoid stuttering. Be careful if modifying this value as we're about to clear pending audio.
-
-		// This is different from the "sb.speaker_enable = false" state
-		// If the speaker is off, this function never gets called
-		// This is speaker on but playing silence
-		soundblaster_mixer_queue.Stop();
-		soundblaster_mixer_queue.Clear();
-		sb.chan->Enable(false);
-	}
-	break;
```

Comment:
Did you move this silence detection elsewhere or just remove it?  It's kind of important for games that primarily use MIDI sound and infrequently play DMA sound effects.  While the SoundBlaster channel is running, it forces the entire mixer into "synchronous mode" where it must wait on the main thread to generate sound.  Allowing the SoundBlaster channel to stop lets the mixer go async with glitch free MIDI.

Reply:
Yep, reverting to glitch-free MIDI is super important. Put that SB DMA to sleep and get it out of the way when we can 😆

Reply:
Oh, it was a comment in the commit that removed it:

> 5. The local 5 second sleep was another Sound Blaster-specific feature
   that the other devices weren't using.  It's now replaced with the
   mixer's 'Sleep' feature.